### PR TITLE
ContactPoseFactor for flat foot constraint

### PR DESCRIPTION
**Pull request by varunagrawal**
This PR adds a new `ContactPoseFactor` which constraints a link to a reference frame (aka `Pose3`), in contrast to `ContactPointFactor` which constrains to a `Point3`. This is very useful for humanoids for both feet as well as grasping contact.

### CODE REVIEW

**Comment by dellaert** [comment]
I'm not going to be able to review. Please enlist help from others while I'm OOO

File: gtdynamics/factors/ContactPointFactor.h [hunk 1]
```diff
@@ -112,4 +112,96 @@ class ContactPointFactor
...
+
+  // The contact point reference frame in the link's CoM frame.
+  gtsam::Pose3 comTcontact_;
+
+ public:
+  /**
+   * Constructor.
+   *
+   * @param link_pose_key Key for the CoM pose of the link in contact.
+   * @param point_key Key for the contact point in the environment.
```
Comment (danbarla) [inline gtdynamics/factors/ContactPointFactor.h:136]: I think a better name here is contact_pose_key, and the description should be "key for the contact point pose in the world/spatial frame"

File: gtdynamics/factors/ContactPointFactor.h [hunk 2]
```diff
@@ -112,4 +112,96 @@ class ContactPointFactor
...
+                    const gtsam::noiseModel::Base::shared_ptr &cost_model,
+                    const gtsam::Pose3 &comTcontact)
+      : Base(cost_model, link_pose_key, point_key), comTcontact_(comTcontact) {}
+
+  /**
+   * Convenience constructor which uses PointOnLink.
+   *
+   * @param point_on_link PointOnLink object which encapsulates the link and its
+   * contact point.
+   * @param point_key Key for the contact point in the environment.
```
Comment (danbarla) [inline gtdynamics/factors/ContactPointFactor.h:150]: same here

File: gtdynamics/factors/ContactPointFactor.h [hunk 3]
```diff
@@ -112,4 +112,96 @@ class ContactPointFactor
...
+      const gtsam::Pose3 &wTl, const gtsam::Pose3 &wTcontact,
+      boost::optional<gtsam::Matrix &> H_link = boost::none,
+      boost::optional<gtsam::Matrix &> H_contact = boost::none) const override {
+    gtsam::Pose3 measured_wTcontact = wTl.compose(comTcontact_, H_link);
+    gtsam::Matrix6 H1, H2;
+    gtsam::Vector error =
+        measured_wTcontact.localCoordinates(wTcontact, H1, H2);
+    if (H_link) *H_link = H1 * (*H_link);
+    if (H_contact) *H_contact = H2;
+    return error;
```
Comment (danbarla) [inline gtdynamics/factors/ContactPointFactor.h:180]: This makes sense to me :)

**Review by danbarla — APPROVED** [review]
This is great, I think it's needed. Just some docs comments.

File: gtdynamics/factors/ContactPointFactor.h (hunk 1 above)
Comment (varunagrawal) [reply gtdynamics/factors/ContactPointFactor.h:136]: Awesome, I was wondering about that too. Glad you could confirm my doubts Dan 😊

**Pull request merged**